Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/GameFileCache.java

```diff
@@ -28,67 +22,89 @@ public GameFileCache(String path)
+    LinkedHashSet<String> pathSet = getPathSet(false);
+    int totalISOPaths =
+            dolphinIni.getInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS, 0);
 
-    if (folderPaths == null)
+    if (!pathSet.contains(path))
     {
-      return;
+      dolphinIni.setInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS,
+              totalISOPaths + 1);
+      dolphinIni.setString(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATH_BASE +
+              totalISOPaths, path);
+      dolphinIni.save(dolphinFile);
     }
-
-    Set<String> newFolderPaths = new HashSet<>(folderPaths);
-    newFolderPaths.add(path);
-    SharedPreferences.Editor editor = preferences.edit();
-    editor.putStringSet(GAME_FOLDER_PATHS_PREFERENCE, newFolderPaths);
-    editor.apply();
   }
 
-  private void removeNonExistentGameFolders(Context context)
+  private static LinkedHashSet<String> getPathSet(boolean removeNonExistentFolders)
   {
-    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
-    Set<String> folderPaths = preferences.getStringSet(GAME_FOLDER_PATHS_PREFERENCE, EMPTY_SET);
+    File dolphinFile = SettingsFile.getSettingsFile(Settings.FILE_DOLPHIN);
+    IniFile dolphinIni = new IniFile(dolphinFile);
+    LinkedHashSet<String> pathSet = new LinkedHashSet<>();
+    int totalISOPaths =
+            dolphinIni.getInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS, 0);
 
-    if (folderPaths == null)
+    if (totalISOPaths < 1)
     {
-      return;
+      return pathSet;
     }
 
-    Set<String> newFolderPaths = new HashSet<>();
-    for (String folderPath : folderPaths)
+    for (int i = 0; i < totalISOPaths; i++)
     {
-      File folder = new File(folderPath);
+      String path = dolphinIni.getString(Settings.SECTION_INI_GENERAL,
+              SettingsFile.KEY_ISO_PATH_BASE + i, "");
+
+      File folder = new File(path);
       if (folder.exists())
       {
-        newFolderPaths.add(folderPath);
+        pathSet.add(path);
       }
     }
 
-    if (folderPaths.size() != newFolderPaths.size())
+    if (removeNonExistentFolders && totalISOPaths > pathSet.size())
     {
-      // One or more folders are being deleted
-      SharedPreferences.Editor editor = preferences.edit();
-      editor.putStringSet(GAME_FOLDER_PATHS_PREFERENCE, newFolderPaths);
-      editor.apply();
+      int setIndex = 0;
+
+      dolphinIni.setInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS,
+              pathSet.size());
+
+      // One or more folders have been removed.
+      for (String entry : pathSet)
+      {
+        dolphinIni.setString(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATH_BASE +
+                setIndex, entry);
+
+        setIndex = setIndex + 1;
+      }
+
+      // Delete known unnecessary keys. Ignore i values beyond index-adjusted totalISOPaths.
+      for (int i = totalISOPaths - pathSet.size() - 1; i < totalISOPaths; i++)
```

Comment:
The start value calculation seems incorrect to me. And even if it is correct, I believe it can be replaced with simply `int i = setIndex` (or you could use `setIndex` directly instead of creating a new variable `i`).

Reply:
Good idea. From what I tested (before the removal of the above empty set check), this is equivalent to `setIndex` but I didn't think to use that so I will.